### ISSUE TITLE
Upload release notes JSON (relnotes.k8s.io) to a Google Cloud Bucket

### ISSUE THREAD

**Issue by saschagrunert** [kubernetes/release#1087]
**What would you like to be added**:

We currently vendor the JSON based release notes for relnotes.k8s.io into the repository:
https://github.com/kubernetes-sigs/release-notes/tree/master/src/assets

We could put these assets on a Google Cloud Storage Bucket to give the Release Notes Team a faster way to update the data as well as avoid having the need for an additional PR on update.

**Why is this needed**:

It's not necessarily needed, but would improve the workflow around the notes and reduce the complexity. 

/cc @cartyc @JamesLaverack @justaugustus

**Comment by justaugustus** [comment]
/assign
/priority important-soon

I need to noodle on this a little, since it will involve granting more people access to GCP...

**Comment by cartyc** [comment]
@justaugustus Is this something that could be delegated to a service account or would that too broad and not be granular enough?

**Comment by JamesLaverack** [comment]
This affects the implementation of `krel release-notes` (see https://github.com/kubernetes/release/blob/master/cmd/krel/cmd/release_notes.go).

@puerco and I were thinking about the implementation, and we realised that in the existing flow we run `release-notes`, then run a JS based prettier for the JSON, then PR into k-sigs/release-notes. (See the [role handbook](https://github.com/kubernetes/sig-release/blob/73a8ef875121987a5baad40fd32d3fdf1fa229c1/release-team/role-handbooks/release-notes/patch-notes.md) for more details). We have to run the prettier (there's a build check on k-sigs/release-notes to make sure you have) and so we'd need to invoke NPM from `krel` to do this currently. So uploading to a bucket would probably make this easier. (We can still pretty-print the JSON, but we can do that in Go instead of invoking the NPM one).

**Comment by saschagrunert** [comment]
> This affects the implementation of `krel release-notes` (see https://github.com/kubernetes/release/blob/master/cmd/krel/cmd/release_notes.go).
> 
> @puerco and I were thinking about the implementation, and we realised that in the existing flow we run `release-notes`, then run a JS based prettier for the JSON, then PR into k-sigs/release-notes. (See the [role handbook](https://github.com/kubernetes/sig-release/blob/73a8ef875121987a5baad40fd32d3fdf1fa229c1/release-team/role-handbooks/release-notes/patch-notes.md) for more details). We have to run the prettier (there's a build check on k-sigs/release-notes to make sure you have) and so we'd need to invoke NPM from `krel` to do this currently. So uploading to a bucket would probably make this easier. (We can still pretty-print the JSON, but we can do that in Go instead of invoking the NPM one).

Sounds good to me! Sure the prettier-invocation would go away if the website fetches the sources from a remote location. :+1:

**Comment by saschagrunert** [comment]
I think the main issue is here which bucket we choose and how we would handle granting permissions to the bucket. At least the whole SIG Release Notes subteam and Release Engineering would have a need to access the bucket…

**Comment by saschagrunert** [comment]
/assign @cartyc 

We will start working on this soon :)

**Comment by saschagrunert** [comment]
Hey @cartyc, do you have any update on this one?

**Comment by cartyc** [comment]
Sorry @saschagrunert I didn't get the notification for some reason. My initial test worked great I'm just stuck right now on getting my head wrapped around on typescript async. I have it so the call gets made but the difficulty I bumped into is exporting that var from `assets.ts` .

```
export const data = async () => {
  const assets = await  gapi.load('client', start);
}
```

There might be a better approach and my newness to typescript is showing :)

**Comment by saschagrunert** [comment]
Hey @cartyc thank you for the update! Do you have something like a WIP PR to play around with? My idea is now that we have all the golang release cut logic in place, we could simply create the release notes JSON directly with the release. Then we can save the result with the release archive, for example there:

 https://console.cloud.google.com/storage/browser/kubernetes-release/archive/anago-v1.20.0-alpha.3?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&prefix=&forceOnObjectsSortingFiltering=false

**Comment by cartyc** [comment]
@saschagrunert I do! I just need to clean it up and then I can push the changed to a work in progress branch. I'll try to have that done in the next couple of days. Is there any eta you're working with?

**Comment by saschagrunert** [comment]
> @saschagrunert I do! I just need to clean it up and then I can push the changed to a work in progress branch. I'll try to have that done in the next couple of days. Is there any eta you're working with?

Cool! Every release we cut from now on is creating a release notes JSON blob into the archive:
```
> gsutil ls "gs://kubernetes-release/archive/anago-v1.20.0-rc.0/release-notes.json"
gs://kubernetes-release/archive/anago-v1.20.0-rc.0/release-notes.json
```

Now I'm wondering if we need to pre-fetch something like an index file which contains a list of existing releases… :thinking: Searching the bucket for those JSON files seems to be really slow.

**Comment by saschagrunert** [comment]
Working on the index file support now in https://github.com/kubernetes/release/pull/1812

**Comment by saschagrunert** [comment]
My target is to finish the process implementation within the next days that we can start consuming the new artifacts with v1.20.0, v1.19.5, v1.18.13 and v1.17.15.

**Comment by cartyc** [comment]
Hey @saschagrunert I opened a WIP PR https://github.com/kubernetes-sigs/release-notes/pull/215

**Comment by saschagrunert** [comment]
> Hey @saschagrunert I opened a WIP PR [kubernetes-sigs/release-notes#215](https://github.com/kubernetes-sigs/release-notes/pull/215)

Thank you! :pray: I'll give it a first quick review.

**Comment by saschagrunert** [comment]
Updating the website to consume the artifacts seems to be the final step before closing this.

**Comment by justaugustus** [comment]
/unassign
/assign @saschagrunert 
FYI @wilsonehusin

**Comment by ashnehete** [comment]
It looks like the implementation on the website side as stalled due to some errors. After debugging @cartyc's PR (https://github.com/kubernetes-sigs/release-notes/pull/215) it seems we could take a different approach and directly fetch `release-notes-index.json` using a public URL i.e. https://gcsweb.k8s.io/gcs/kubernetes-release/release/release-notes-index.json

The only blocker here seems that the above URL doesn't support CORS which is required to make AJAX calls from [relnotes.k8s.io](https://relnotes.k8s.io).

/assign @ashnehete

**Comment by xmudrii** [comment]
This is blocked by https://github.com/kubernetes/k8s.io/issues/1810#issuecomment-804524844 until we don't move to the community GCP project.

**Comment by ashnehete** [comment]
I just realised that [cdn.dl.k8s.io](https://cdn.dl.k8s.io/release/release-notes-index.json) is enabled for CORS which means we can fetch release assets using this URL. Moreover, using the [dl.k8s.io](dl.k8s.io) domain seems to be where the community is moving as a whole too.

I'm currently working on a PR for this in the release notes repo.

/lifecycle active

**Comment by ashnehete** [comment]
Now that the PR is merged and site is actively fetching notes from the remote, do we need to amend the release-notes team docs to stop them from creating a PR for each new release? Any code changes needed in the `krel` utility?

**Comment by saschagrunert** [comment]
> Now that the PR is merged and site is actively fetching notes from the remote, do we need to amend the release-notes team docs to stop them from creating a PR for each new release? Any code changes needed in the `krel` utility?

Hm I don't think so, the release notes should be now populated automatically. :thinking:
